torch/testing/_internal/common_methods_invocations.py

```diff
@@ -4547,6 +4547,22 @@ def _tensor(shape, dtype=dtype, low=None, high=None):
 
     return [SampleInput(tensor, args=args) for tensor, args in test_cases]
 
+def sample_inputs_tensorinv(op_info, device, dtype, requires_grad, **kwargs):
+    def make_input():
+        input = make_fullrank_matrices_with_distinct_singular_values(12, 12, device=device, dtype=dtype)
+        return input.requires_grad_(requires_grad)
+
+    # lhs / rhs shape can have any number of dimensions as long as their product equals 12
+    shapes = [
+        ((2, 2, 3), (12, 1)),
+        ((4, 3), (6, 1, 2)),
+    ]
+
+    return [
+        SampleInput(make_input().reshape(*shape_lhs, *shape_rhs), kwargs=dict(ind=len(shape_lhs)))
+        for shape_lhs, shape_rhs in shapes
+    ]
```

Comment:
Nit. You might want to consider having a generator for readability. See for example https://github.com/pytorch/pytorch/blob/1c113d227ca42c5af31f5f87d801c94f51b2b82e/torch/testing/_internal/common_methods_invocations.py#L852-L856
In an ideal world, `OpInfo`s would return iterators and we could potentially speed things up / reduce the memory footprint.

Reply:
> In an ideal world, `OpInfo`s would return iterators and we could potentially speed things up / reduce the memory footprint.

I can get behind that, but I fail to see how the approach you suggested is easier. Let's walk through the possible cases.

We start with the current state, i.e. returning a list of samples:

1. Simple creation logic without any branching:

```python
return [
    SampleInput(...)
    for args in args_cases
]
```

2. Complicated creation logic with branching

```python
sample_inputs = []
for args in args_cases:
    # complicated branching logic
    sample_inputs.append(SampleInput(...))
```

Both cases are easy to read, concise, and do not require an extra function. So IMO they are superior to the proposed `generator` function that gets converted into a list unconditionally.

Let's have a look at the future scenario where we want to return generators over lists. What would we need to change if we go without the `generator` design?

1. Replace `[]` by `()`

```python
return (
    SampleInput(...)
    for args in args_cases
)
```

2. `yield` instead of appending to a list and return it

```python
for args in args_cases:
    # complicated branching logic
    yield SampleInput(...)
```

So in both cases it as easy to go from a returned list to a generator as removing the `list` call from `return list(generator())`.

Maybe I'm missing something, but so far I fail to see how your suggested approach is superior to mine. Happy to be enlightened.

Reply:
Sorry, I said that "In an ideal world, OpInfos would return iterators", but what I wanted to say is that "OpInfo samplers would be generators", meaning that you'd write:

```python
def sample_inputs_foo(op_info, device, dtype, requires_grad):
  sizes = ...  
  for size in sizes:
    yield SampleInput(...) 
```

Note that in this case, all the generators, simple or complex, would have the same structure. We would not need to change between list comprehensions / generator expressions and the for loop syntax depending on the complexity of the generation (note also that the generator expression makes the strong assumption that the returned iterable is just iterated over once!). I also personally find much cleaner a `yield` than a `sample_inputs.append()` where you have to declare an empty list at the beginning and return it at the end. You also need to write `sample_inputs: list[SampleInput] = []` to make mypy happy, which is annoying. 

The other point of letting samplers being generators is that, if you just need to use each sample once, you can run the tests on every sample and each sample is deleted after being processed, rather than having to create all the samples upfront and having to keep them in memory for all the duration of the test.

Reply:
> (note also that the generator expression makes the strong assumption that the returned iterable is just iterated over once!)

How is this different in the "ideal world"? If you want to iterate multiple times of the sample inputs, you would still need to call the function multiple times. AFAIK there is no functional difference between:

```python
def foo():
    for bar in baz:
        yield bar
```

and 


```python
def foo():
    return (bar for bar in baz)
```

> You also need to write `sample_inputs: list[SampleInput] = []` to make mypy happy, which is annoying.

Nope. Unless you are adding anything else besides a `SampleInput` to the list, `mypy` figures this out automatically. You only need to annotate empty containers in case you add multiple types to it.

> The other point of letting samplers being generators is that, if you just need to use each sample once, you can run the tests on every sample and each sample is deleted after being processed, rather than having to create all the samples upfront and having to keep them in memory for all the duration of the test.

Yeah, that is true. But in the current state neither my nor the approach you linked will do that. Since I've shown that my approach can be converted to a generator approach as easy the linked one, I fail to see how this is an argument for changing mine.

Reply:
You are right, both implementations of `foo` above are equivalent. My mistake.

And about mypy, I remember having to do this at some point, but I don't really know.

And yes, my comment was not so much about this particular code but more general. Nik, Kshiteej and me started using the "generator pattern" (writing a generator and then returning `list(gen())`) to potentially have an easier transition to generators later on once they are supported. Perhaps the missing link here is that no one actually sat down and adapted the code of the tests for it to accept generators :D

Then again, this was more of a comment than anything feel free to resolve :)